src/ppx/scene/scene_node.cpp

```diff
@@ -0,0 +1,250 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "ppx/scene/scene_node.h"
+
+namespace ppx {
+namespace scene {
+
+// -------------------------------------------------------------------------------------------------
+// Node
+// -------------------------------------------------------------------------------------------------
+Node::Node(scene::Scene* pScene)
+    : mScene(pScene)
+{
+}
+
+Node::~Node()
+{
+}
+
+void Node::SetVisible(bool visible)
+{
+    mVisible = visible;
+}
+
+const float4x4& Node::GetEvaluatedMatrix() const
+{
+    if (mEvaluatedDirty) {
+        float4x4 parentEvaluatedMatrix = float4x4(1);
+        if (!IsNull(mParent)) {
+            parentEvaluatedMatrix = mParent->GetEvaluatedMatrix();
+        }
+        auto& concatenatedMatrix = GetConcatenatedMatrix();
+        mEvaluatedMatrix         = parentEvaluatedMatrix * concatenatedMatrix;
+        mEvaluatedDirty          = false;
+    }
+    return mEvaluatedMatrix;
+}
+
+void Node::SetParent(scene::Node* pNewParent)
+{
+    mParent = pNewParent;
+}
+
+void Node::SetEvaluatedDirty()
+{
+    mEvaluatedDirty = true;
+    for (auto& pChild : mChildren) {
+        pChild->SetEvaluatedDirty();
+    }
+}
+
+void Node::SetTranslation(const float3& translation)
+{
+    Transform::SetTranslation(translation);
+    SetEvaluatedDirty();
+}
+
+void Node::SetRotation(const float3& rotation)
+{
+    Transform::SetRotation(rotation);
+    SetEvaluatedDirty();
+}
+
+void Node::SetScale(const float3& scale)
+{
+    Transform::SetScale(scale);
+    SetEvaluatedDirty();
+}
+
+void Node::SetRotationOrder(Transform::RotationOrder value)
+{
+    Transform::SetRotationOrder(value);
+    SetEvaluatedDirty();
+}
+
+scene::Node* Node::GetChild(uint32_t index) const
+{
+    scene::Node* pChild = nullptr;
+    if (ppx::GetElement(index, mChildren, &pChild)) {
+        return nullptr;
+    }
+    return pChild;
+}
+
+bool Node::IsInSubTree(const scene::Node* pNode)
+{
+    bool inSubTree = (pNode == this);
+    if (!inSubTree) {
+        for (auto pChild : mChildren) {
+            inSubTree = pChild->IsInSubTree(pNode);
+            if (inSubTree) {
+                break;
+            }
+        }
+    }
+    return inSubTree;
+}
+
+ppx::Result Node::AddChild(scene::Node* pNewChild)
+{
+    // Cannot add child if current node is standalone
+    if (IsNull(mScene)) {
+        return ppx::ERROR_SCENE_INVALID_STANDALONE_OPERATION;
+    }
+
+    // Cannot add NULL child
+    if (IsNull(pNewChild)) {
+        return ppx::ERROR_UNEXPECTED_NULL_ARGUMENT;
+    }
+
+    // Cannot add self as a child
+    if (pNewChild == this) {
+        return ppx::ERROR_SCENE_INVALID_NODE_HIERARCHY;
+    }
+
+    // Cannot add child if current node is in child's subtree
+    if (IsInSubTree(this)) {
```

Comment:
Should this be `pNewChild->IsInSubTree(this)`?

Reply:
Yep. Fixed.